saul-core/src/main/java/edu/illinois/cs/cogcomp/saul/learn/MultilayerPerceptron.java

```diff
@@ -0,0 +1,583 @@
+        super("");
+        int[] layer_size = new int[hidden_layers.length+2];
+        layer_size[0] = num_inputs;
+        for (int i = 0; i < hidden_layers.length; i++){
+            layer_size[i+1] = hidden_layers[i];
+        }
+        layer_size[hidden_layers.length+1] = num_outputs;
+
+        layers = new Layer[hidden_layers.length+1];
+        for (int i = 0; i < layers.length; i++){
+            layers[i] = new Layer(layer_size[i],layer_size[i+1],activation_function[i]);
+        }
+    }
+
+    /**
+    * Trains the neural network on a given example, given a real input and output vector, using default learning parameters.
+    *
+    * @param input      Real vector consisting of the input features of an example
+    * @param output     Real vector consisting of the labelled outputs of the neural network.
+    */
+    public void learn(double[] input, double[] output){
+        learn(input,output,default_learning_rate,default_lambda,default_momentum);
+    }
+
+    /**
+     * Trains the neural network on a given example, given a real input and output vector, using given learning parameters.
+     *
+     * @param input         Real vector consisting of the input features of an example
+     * @param output        Real vector consisting of the labelled outputs of the neural network.
+     * @param learning_rate Learning rate
+     * @param lambda        Penalization constant for L-2 regularization
+     * @param momentum      Momentum value for speeding convergence
+     */
+    public void learn(double[] input, double[] output, double learning_rate, double lambda, double momentum){
+        reset_errors();
+        compute_activation(input);
+        output_errors(output);
+        compute_gradients(input);
+        update_weights(learning_rate,lambda, momentum);
+    }
+
+    /**
+     * Trains the neural network on a sparse input vector, using default learning parameters.
+     *
+     * @param exampleFeatures The example's array of feature indices.
+     * @param exampleValues The example's array of feature values.
+     * @param exampleLabels The example's label(s). Should be {0} or {1}.
+     * @param labelValues The labels' values.
+     */
+    public void learn(int[] exampleFeatures, double[] exampleValues, int[] exampleLabels, double[] labelValues){
+        if (((++iterations)%10000) == 0) {
+            System.out.println("learning: " + iterations);
+        }
+        assert exampleLabels.length == 1 : "Example must have a single label.";
+        double[] output = new double[layers[layers.length-1].errors.length];
+        Arrays.fill(output,0.0d);
+        output[exampleLabels[0]] = 1.0d;
+        int maximum_index = -1;
+        for (int i = 0; i < exampleFeatures.length; i++){
+            if  (exampleFeatures[i] > maximum_index){
+                maximum_index = exampleFeatures[i];
+            }
+        }
+        if (layers[0].weights.length < maximum_index + 2 )
+            resize_inputs(maximum_index + 10000);
```

Comment:
For later: It might be better to do a pre-extract step and calculate the final label lexicon size. This might add a high computation overhead if we do not have 10k more features.